### ISSUE TITLE
Rule proposal: `use-baseline` to warn when using non-baseline APIs

### ISSUE THREAD

**Issue by guillaumebrunerie** [typescript-eslint/typescript-eslint#11577]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

TypeScript's policy is that as soon as a JavaScript/DOM API is available in two major implementations, it is marked as being unconditionally present. This may cause problems with for instance code that ends up crashing in Safari because the API is only available in Chrome and Firefox, but TypeScript does not warn about it and the developer did not test in Safari.

The new CSS linting released by ESLint includes a `use-baseline` rule (https://github.com/eslint/css/blob/main/docs/rules/use-baseline.md), which I think would be great to have for JavaScript/TypeScript as well. It probably belong here rather that in ESLint core as it would need type information to work reliably.

### Fail Cases

```TypeScript
navigator.vibrate();

document.startTransition(...);

requestIdleCallback(...);

const parent = document.getElementById("parent");
const element1 = document.getElementById("element1");
const element2 = document.getElementById("element2");
parent.moveBefore(element1, element2); // <- here, needs type information
```

### Pass Cases

```TypeScript
if (navigator.vibrate) {
  navigator.vibrate();
}

const doSomething = () => {
  if (!document.startTransition) return;
  document.startTransition(...)
}

requestIdleCallback?.(...)
```

### Additional Info

It should be possible to specify which Baseline is acceptable (maybe in the same way as the `use-baseline` rule in ESLint CSS, for consistency), and also to disable the warning for specific APIs (for instance if you include a polyfill for some API it should be OK to use it unconditionally anywhere).

I’m not sure how easy it is to make the "Pass Cases" above pass, maybe it would need some complicated control flow analysis?

There is already a plugin for this (`eslint-plugin-compat`), but it is pretty broken in practice (see for instance https://github.com/amilajack/eslint-plugin-compat/issues/404, anything inside any `if` statement is exempt), and I feel like it should be in typescript-eslint core instead (just like it is in ESLint for CSS).

See also the discussion at https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1486.

**Comment by bradzacher** [comment]
I don't think this should be a rule we maintain because then we'd have to build a bunch of infra to pulll and reference the caniuse datasets or scrape MDN or whatever. There are also so many combinations of things that it would be burdensome to test.

Setting rules on how guarding should work -- or even just trying to check for valid guards of specific features...

There's a huge chunk of work here -- and so I don't think it's something we should pickup -- we barely  have the bandwidth to maintain our project as it is.

IMO this should be its own plugin in of itself -- and it does, right? But the fact that nobody's maintaining it shows that either the community drive for this is low and does not outweigh the maintenance cost.

Thoughts @typescript-eslint/triage-team ?

---

My personal take:
Use your package manager's builtin patch tooling to patch the TS definitions to mark the relevant APIs as optional (or delete them entirely). I mean ultimately https://github.com/microsoft/TypeScript-DOM-lib-generator/ this is open source and so you could use it yourself to generate your own packages and leverage the `@typescript/lib-xxx` tooling built in to TS to swap-out the default types with your own crafted to your browser version.

**Comment by JoshuaKGoldberg** [comment]
Yeah, I think this would be a useful rule in its own plugin - but we're not well-positioned to take it on.

**Comment by JoshuaKGoldberg** [comment]
Closing out as we have at a majority of team members agreeing this belongs elsewhere. If you're interested in building this standalone, see:

* https://typescript-eslint.io/developers/custom-rules for building custom rules
* https://typescript-eslint.io/developers/eslint-plugins for publishing it as a third-party plugin

Cheers! 💙
